File: Temp/Java/Practice/src/sort/QuickSort.java
package sort;

public class QuickSort {
	private int[] arr;
	
	public QuickSort(int[] arr) {
		this.arr = arr;
	}
	
	public void solve() {
		System.out.println("Quick Sort:");
		quickSort(0, arr.length - 1);
	}
	
	void quickSort(int lowereBound, int upperBound) {
		if (lowereBound < upperBound) {
			int pivotIndex = partition(lowereBound, upperBound);
			
			quickSort(lowereBound, pivotIndex - 1);
			quickSort(pivotIndex + 1, upperBound);
		}
	}

	private int partition(int lowerBound, int upperBound) {
		int pivot = arr[upperBound];
		int lowerBoundIndex = lowerBound;
		
		for (int index = lowerBoundIndex; index < upperBound; index++)
			if (arr[index] < pivot)
				swap(index, lowerBoundIndex++);
		swap(lowerBoundIndex, upperBound);
		return lowerBoundIndex;
	}

	private void swap(int indexA, int indexB) {
		int tmp = arr[indexA];
		arr[indexA] = arr[indexB];
		arr[indexB] = tmp;
	}
	
}

File: Temp/Python/Practice/Sort/QuickSort.py
class QuickSort(object):
    def __init__(self, arr=[]):
        super(QuickSort, self).__init__()
        self.arr = arr

    def solve(self):
        self.__quick_sort(self.arr, 0, len(self.arr) - 1)

    def __quick_sort(self, arr, lower_bound, upper_bound):
        if lower_bound < upper_bound:
            pivot_index = self.__partition(arr, lower_bound, upper_bound)

            self.__quick_sort(arr, lower_bound, pivot_index - 1)
            self.__quick_sort(arr, pivot_index + 1, upper_bound)

    @staticmethod
    def __partition(arr, lower_bound, upper_bound):
        pivot = arr[upper_bound]
        lth = lower_bound
        for ith in range(lth, upper_bound):
            if arr[ith] < pivot:
                arr[ith], arr[lth] = arr[lth], arr[ith]
                lth += 1
        arr[lth], arr[upper_bound] = arr[upper_bound], arr[lth]
        return lth


def get_instance(arr=[]):
    return QuickSort(arr=arr)
